### ISSUE TITLE
`Method doesn't allow unregistered callers (callers without established identity). Please use API Key or other form of API consumer identity to call this API.` with no-key 

### ISSUE THREAD

**Issue by Benjamin-Loison** [Benjamin-Loison/YouTube-operational-API#237]
```json
{
  "error": {
    "code": 403,
    "message": "Method doesn't allow unregistered callers (callers without established identity). Please use API Key or other form of API consumer identity to call this API.",
    "errors": [
      {
        "message": "Method doesn't allow unregistered callers (callers without established identity). Please use API Key or other form of API consumer identity to call this API.",
        "domain": "global",
        "reason": "forbidden"
      }
    ],
    "status": "PERMISSION_DENIED"
  }
}
```

Related to #229 (hence also [this initial Git issue comment](https://github.com/Benjamin-Loison/YouTube-operational-API/issues/11#issuecomment-1882945277)).

**Comment by Benjamin-Loison** [comment]
```python
#!/usr/bin/python3

import requests
from tqdm import tqdm
import json

with open('ytPrivate/keys.txt') as f:
    lines = f.read().splitlines()

for line in tqdm(lines):
    #print(line)
    url = 'https://www.googleapis.com/youtube/v3/videos'
    params = {
        'part': 'snippet',
        'id': '_ZPpU7774DQ',
        'key': line,
    }
    data = requests.get(url, params = params).json()
    if data['error']['errors'][0]['reason'] != 'quotaExceeded':
        print(json.dumps(data, indent = 4))
        break
```

with https://yt.lemnoslife.com key set shows that no key still have any quota.

Note that I ran this script in 18 seconds on my Linux Mint Framework and I tested manually a few keys in Firefox and noticed only quota being exceeded.

Are all keys somehow empty, if they can even be or there was more quota usage than usually? I should monitor key usage.

In the meanwhile I should look for new keys.

**Comment by Benjamin-Loison** [comment]
As below script run under 16 seconds on OC3K, I programmed a crontab to run every minute to let me know once a key has quota.

```python
#!/usr/bin/python3

import requests
import json

with open('/var/www/yt/ytPrivate/keys.txt') as f:
    lines = f.read().splitlines()

for line in lines:
    url = 'https://www.googleapis.com/youtube/v3/videos'
    params = {
        'part': 'snippet',
        'id': '_ZPpU7774DQ',
        'key': line,
    }
    data = requests.get(url, params = params).json()
    if data['error']['errors'][0]['reason'] != 'quotaExceeded':
        print(json.dumps(data, indent = 4))
        break
```

**Comment by Benjamin-Loison** [comment]
#91 would make the no-key endpoint and its maintenance no more important.

**Comment by Benjamin-Loison** [comment]
On https://yt0.lemnoslife.com I noticed in `/var/log/apache2/` after a reboot:

```
==> error.log <==
[Thu Feb 15 23:27:59.543673 2024] [mpm_prefork:error] [pid 1023] AH00161: server reached MaxRequestWorkers setting, consider raising the MaxRequestWorkers setting
```

`yt.lemnoslife.com-ssl--access.log` seems to show many requests to its no-key endpoint for https://yt0.lemnoslife.com.

**Comment by Benjamin-Loison** [comment]
Manually merging:
- https://yt.lemnoslife.com
- https://yt0.lemnoslife.com
- https://yt4.lemnoslife.com

**Comment by Benjamin-Loison** [comment]
Consider #18.

**Comment by Benjamin-Loison** [comment]
Unclear what was current assumption but let us now keep all keys even those that are quota exceeded, assuming that no key can have no quota forever.

**Comment by Benjamin-Loison** [comment]
With new `keys.txt`:

```
correct=2056 quotaExceeded=1571: 100%|█████████████████████████████████████████████████████████████████████████████████████| 3627/3627 [05:56<00:00, 10.19it/s]
```

Previously had about 268 keys.

**Comment by Benjamin-Loison** [comment]
**TLDR: incident solved (for the moment at least), you can use the no-key endpoint as previously**
Previously the no-key endpoint was empowered by about 262 YouTube Data API v3 keys that I mostly found by hand on the Internet. However, as I said in the last hours all keys were having their quota exceeded. Hence, I wrote web-scraping algorithms to automatize my Internet search for YouTube Data API v3 keys and I found a very efficient source which resulted in 3,627 keys with 2,056 not currently having their quota expired. I updated the official and private instances key set with the one I elaborated. Except if Google is scrutinizing my work, we should no more have any quota related issue concerning the no-key endpoint for a long time.

**Comment by Benjamin-Loison** [comment]
Note that for unknown reason https://yt0.lemnoslife.com no-key endpoint is not working but as it is not officially supported let us do not care about it for the moment.

**Comment by Benjamin-Loison** [comment]
Related to [Benjamin-Loison/matrix-commander/issues/8](https://github.com/Benjamin-Loison/matrix-commander/issues/8).

**Comment by Benjamin-Loison** [comment]
Another possibility to get YouTube Data API v3 could be to get those of big companies, thanks to [this method](https://discord.com/channels/933841502155706418/933841503103627316/1208189013404549151).

**Comment by Benjamin-Loison** [comment]
Just got [this error](https://github.com/Benjamin-Loison/YouTube-operational-API/issues/237#issue-2135183347), the second line of `keys.txt` was empty (or whitespaces only), I solved the issue by removing this empty line thanks to a Linux command.

**Comment by Benjamin-Loison** [comment]
```py
import requests
from tqdm import tqdm

ytUrl = 'https://www.googleapis.com/youtube/v3/videos'
ytParams = {
    'part': 'snippet',
    'id': '_ZPpU7774DQ',
}

correctKeys = []
quotaExceededKeys = []

def getValue(obj, path):
    if path == '':
        return obj
    pathParts = path.split('/')
    key = pathParts[0]
    key = int(key) if key.isdigit() else key
    try:
        newObj = obj[key]
        return getValue(newObj, '/'.join(pathParts[1:]))
    except:
        return None

with open('b') as f:
    keys = f.read().splitlines()
    for key in tqdm(keys):
        ytParams['key'] = key
        data = requests.get(ytUrl, params = ytParams).json()
        if getValue(data, 'items/0/snippet/channelId') == 'UCWeg2Pkate69NFdBeuRFTAw':
            correctKeys += [key]
            print(f'{len(correctKeys)=}')
        elif data['error']['errors'][0]['reason'] == 'quotaExceeded':
            quotaExceededKeys += [key]
            print(f'{len(quotaExceededKeys)=}')

print('\n'.join(quotaExceededKeys + correctKeys))
```
